cmd/snap/inhibit.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/progress"
+)
+
+func inhibitMessage(snapName string, hint runinhibit.Hint) string {
+	switch hint {
+	case runinhibit.HintInhibitedForRefresh:
+		return fmt.Sprintf(i18n.G("snap package %q is being refreshed, please wait"), snapName)
+	default:
+		return fmt.Sprintf(i18n.G("snap package cannot be used now: %s"), string(hint))
+	}
+}
+
+func isGraphicalSession() bool {
+	return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
+}
+
+var hasZenityExecutable = func() bool {
+	return osutil.ExecutableExists("zenity")
+}
+
+func zenityFlow(snapName string, hint runinhibit.Hint) error {
+	zenityTitle := i18n.G("Snap package cannot be used")
+
+	// Run zenity with a progress bar.
+	// TODO: while we are waiting ask snapd for progress updates and send those
+	// to zenity via stdin.
+	zenityDied := make(chan error, 1)
+
+	// TODO: use a dbus API to allow integration with native desktop environment.
+	cmd := exec.Command(
+		"zenity",
+		// [generic options]
+		"--title="+zenityTitle,
+		// [progress options]
+		"--progress",
+		"--text="+inhibitMessage(snapName, hint),
+		"--pulsate",
+	)
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	// Make sure that zenity is eventually terminated.
+	defer cmd.Process.Signal(os.Interrupt)
+	// Wait for zenity to terminate and store the error code.
+	// The way we invoke zenity --progress makes it wait forever.
+	// so it will typically be an external operation.
+	go func() {
+		zenityErr := cmd.Wait()
+		if zenityErr != nil {
+			zenityErr = fmt.Errorf("zenity error: %s\n", zenityErr)
+		}
+		zenityDied <- zenityErr
+	}()
+
+	if err := waitInhibitUnlock(snapName, zenityDied); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func textFlow(snapName string, hint runinhibit.Hint) error {
+	fmt.Fprintf(Stdout, "%s\n", inhibitMessage(snapName, hint))
+	pb := progress.MakeProgressBar()
+	pb.Spin(i18n.G("please wait..."))
+	err := waitInhibitUnlock(snapName, nil)
+	pb.Finished()
+	return err
+}
+
+func waitWhileInhibited(snapName string) error {
+	hint, err := runinhibit.IsLocked(snapName)
+	if err != nil {
+		return err
+	}
+	if hint == runinhibit.HintNotInhibited {
+		return nil
+	}
+
+	if isGraphicalSession() && hasZenityExecutable() {
+		return zenityFlow(snapName, hint)
+	}
+	// terminal and headless
+	return textFlow(snapName, hint)
+}
+
+var isLocked = runinhibit.IsLocked
+
+var waitInhibitUnlock = func(snapName string, errCh <-chan error) error {
+	// Every second check if the inhibition file is still present.
+	ticker := time.NewTicker(500 * time.Millisecond)
+	defer ticker.Stop()
+loop:
+	for {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				fmt.Fprintf(Stderr, "%s", err)
+			}
+			break loop
+		case <-ticker.C:
+			// A second has elapsed, let's check again.
+			hint, err := isLocked(snapName)
+			if err != nil {
+				return err
+			}
+			if hint == runinhibit.HintNotInhibited {
```

Comment:
now that we have  HintInhibitedGateRefresh  we probably need more complex behaviors, we probably want to wait without displaying anything while the lock is HintInhibitedGateRefresh and display things only once it's really HintInhibitedForRefresh

Reply:
Done by waiting for InhibitedForRefresh first in case it is InhibitedGateRefresh (after checking it's not HintNotInhibited first), then waiting for it to become NotInhibited.
But having second thoughts, I wonder if we need to explicitly wait for HintInhibitedForRefresh.

Reply:
are we avoiding displaying anything until  InhibitedForRefresh ? that was the point of asking to split the wait, otherwise indeed I suppose is not needed

Reply:
Yes, I made changes to waitWhileInhibited not to display antything until InhibitedForRefresh.

Reply:
> we probably want to wait without displaying anything while the lock is HintInhibitedGateRefresh 

@pedronis but why though? as Pawel pointed out, it could take an arbitrary amount of time for that hook to finish and proceed to HintInhibitedForRefresh in which case the user sees nothing, if we refactored the flow (I guess perhaps how the flow was originally ?) then we could display something that blocks waiting for both cases to change and thus have less cases of user confusion wondering why nothing is happening when they run their snap

Reply:
I'll comment on this. I understand the concerns. I still think this is the baseline behavior we want. On which can improve later as needed.